### DISCUSSION TITLE
How to use AI Deck

### DISCUSSION THREAD

**Discussion by akshaysunkara** [bitcraze/discussions#1434] [Q&A]
Hey guys, I'm trying to use the AI deck to obtain camera footage from the drone.

I think I attached the AI deck properly, however, I don't see it whenever I'm on cfclient gui. 

![image](https://github.com/bitcraze/discussions/assets/171630750/34d6e534-4476-4f41-92b3-ae375fdac013)

But in the console, this is the log it gives me:

SYS: ----------------------------
SYS: Crazyflie 2.1 is up and running!
SYS: Build 0:e527895e8c6e (2023.11) CLEAN
SYS: I am 0x2037343548435012002D001A and I have 1024KB of flash!
CFGBLK: v1, verification [OK]
DECK_CORE: 2 deck(s) found
DECK_CORE: Calling INIT on driver bcAI for deck 0
CPX-EXT-ROUTER: Waiting for CPX External router initialization
CPX-EXT-ROUTER: CPX External router initialized, CPX_VERSION: 0
AIDECK: Not setting up WiFi
DECK_CORE: Calling INIT on driver bcFlow2 for deck 1
ZR2: Z-down sensor [OK]
PMW: Motion chip id: 0x49:0xB6
IMU: BMI088: Using I2C interface.
IMU: BMI088 Gyro connection [OK].
IMU: BMI088 Accel connection [OK]
IMU: BMP388 I2C connection [OK]
ESTIMATOR: Using Kalman (2) estimator
CONTROLLER: Using PID (1) controller
MTR-DRV: Using brushed motor driver
SYS: About to run tests in system.c.
SYS: NRF51 version: 2024.2 (CF21)
EEPROM: I2C connection [OK].
STORAGE: Storage check [OK].
IMU: BMI088 gyro self-test [OK]
DECK_CORE: Deck 0 test [OK].
DECK_CORE: Deck 1 test [OK].
SYS: Self test passed!
STAB: Wait for sensor calibration...
SYS: Free heap: 11704 bytes
STAB: Starting stabilizer loop
ESTKALMAN: State out of bounds, resetting
SUP: Can not fly
SUP: Ready to fly

I've attached both the flowdeck and the AI deck. 

So, should I continue with the steps provided on the Getting started page for the AI deck? https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/

Or do I need to fix this before I continue with the steps?

**Comment by knmcguire** [comment]
Hi! Your crazyflie is detecting the AIdeck it seems, so you can continue with the tutorial!

```
DECK_CORE: Calling INIT on driver bcAI for deck 0
...
AIDECK: Not setting up WiFi
```

**Reply by akshaysunkara** [reply]
Hey, I still have a lot of problems while trying to flash the drone with the AI deck:

Whenever, I'm on the step Gap8 Loader and run these commands:

$ git clone https://github.com/bitcraze/aideck-gap8-bootloader.git
$ cd aideck-gap8-bootloader
$ docker run --rm -it -v $PWD:/module/ --device /dev/ttyUSB0 --privileged -P bitcraze/aideck /bin/bash -c 'export GAPY_OPENOCD_CABLE=interface/ftdi/olimex-arm-usb-tiny-h.cfg; source /gap_sdk/configs/ai_deck.sh; cd /module/;  make all image flash'

I get this error: 

The term 'docker' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a 
path was included, verify that the path is correct and try again.
At line:1 char:1

I don't know if I even need to do that step since all I want is to get footage from the AI deck so when I go ahead with the steps on Flash Wifi Example and run this code, I get this error: (I changed the commands in the steps)

PS C:\Users\aksha\Desktop\MIT> python3 crazyflie-clients-python\src\cfloader flash aideck_gap8_wifi_img_streamer_with_ap.bin deck-bcAI:gap8-fw -w radio://0/80/2M/E7E7E7E7E7
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\aksha\Desktop\MIT\crazyflie-clients-python\src\cfloader\__main__.py", line 1, in <module>
    from cfloader import main
ImportError: cannot import name 'main' from 'cfloader' (C:\Users\aksha\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\cfloader\__init__.py)

These are how my folders are:
![image](https://github.com/bitcraze/discussions/assets/171630750/23c3ac71-9f1a-4f37-bff6-9e9389c4d89a)

What command should I run to get this to work? Do I have the follow the Gap8 Flashing thing?

**Reply by akshaysunkara** [reply]
Just to give more information, the cfloader file has these files called __main__ and __init__

This is the code in both of these files:

__main__:
from cfloader import main

if __name__ == "__main__":
    main()

__init__:

import sys

import cflib.crtp
from cflib.bootloader import Bootloader, Target
from cflib.bootloader.boottypes import BootVersion


def main():
    # Initialise the CRTP link driver
    try:
        cflib.crtp.init_drivers()
    except Exception as e:
        print("Error: {}".format(str(e)))
        sys.exit(-1)

    # Set the default parameters
    clink = None
    action = "info"
    boot = "cold"
    filename = None 
    targets = None  
    bl = None  

    exit_result = 0

    if len(sys.argv) < 2:
        print()
        print("==============================")
        print(" CrazyLoader Flash Utility")
        print("==============================")
        print()
        print(" Usage:", sys.argv[0], "[CRTP options] <action> [parameters]")
        print()
        print("The CRTP options are described above")
        print()
        print("Crazyload option:")
        print("   info                    : Print the info of the bootloader "
              "and quit.")
        print("                             Will let the target in bootloader "
              "mode")
        print("   reset                   : Reset the device in firmware mode")
        print("   flash <file> [targets]  : flash the <img> binary file from "
              "the first")
        print("                             possible  page in flash and reset "
              "to firmware")
        print("                             mode.")
        sys.exit(0)

    # Analyse the command line parameters
    sys.argv = sys.argv[1:]
    argv = []

    i = 0
    while i < len(sys.argv):
        if sys.argv[i] == "--cold-boot" or sys.argv[i] == "-c":
            boot = "cold"
        elif sys.argv[i] == "--warm-boot" or sys.argv[i] == "-w":
            boot = "reset"
            i += 1
            clink = sys.argv[i]
        else:
            argv += [sys.argv[i]]
        i += 1
    sys.argv = argv

    # Analyse the command
    if len(sys.argv) < 1:
        action = "info"
    elif sys.argv[0] == "info":
        action = "info"
    elif sys.argv[0] == "reset":
        action = "reset"
    elif sys.argv[0] == "flash":
        if len(sys.argv) < 2:
            print("The flash action require a file name.")
            sys.exit(-1)
        action = "flash"
        filename = sys.argv[1]
        targets = []  # Dict[Target]
        for t in sys.argv[2:]:
            if t.startswith("deck-"):
                [deck, target, type] = t.split("-")
                targets.append(Target("deck", target, type, [], []))
            else:
                [target, type] = t.split("-")
                targets.append(Target("cf2", target, type, [], []))
    else:
        print("Action", sys.argv[0], "unknown!")
        sys.exit(-1)

    try:
        # Initialise the bootloader lib
        bl = Bootloader(clink)

        warm_boot = (boot == "reset")
        if warm_boot:
            print("Reset to bootloader mode ...")
            sys.stdout.flush()
        else:  # The connection is done by a cold boot ...
            print("Restart the Crazyflie you want to bootload in the next"),
            print(" 10 seconds ..."),

            sys.stdout.flush()

        ######################################
        # Doing something (hopefully) useful
        ######################################

        if action == "info":
            def print_info(version: int, connected_targets: [Target]):
                print("Connected to bootloader on {} (version=0x{:X})".format(
                    BootVersion.to_ver_string(version),
                    version
                    )
                )
                for target in connected_targets:
                    print(target)

            # flash_full called with no filename will not flash, just call
            # our info callback
            bl.flash_full(None, None, warm_boot, None, print_info)
        elif action == "flash" and filename:
            is_target_required = not filename.endswith('.zip')
            if (is_target_required and not targets):
                print("The flash action with a non .zip file requires a target")
                sys.exit(-1)
            try:
                bl.flash_full(None, filename, warm_boot, targets)
            except Exception as e:
                print("Failed to flash: {}".format(e))
                exit_result = -1
        elif action == "reset":
            bl.reset_to_firmware()
        else:
            None
    except Exception as e:
        import traceback

        traceback.print_exc(file=sys.stdout)
        print(e)
        exit_result = -1

    finally:
        #########################
        # Closing the connection
        #########################
        if bl:
            bl.close()

    sys.exit(exit_result)

import sys

import cflib.crtp
from cflib.bootloader import Bootloader, Target
from cflib.bootloader.boottypes import BootVersion


def main():
    # Initialise the CRTP link driver
    try:
        cflib.crtp.init_drivers()
    except Exception as e:
        print("Error: {}".format(str(e)))
        sys.exit(-1)

    # Set the default parameters
    clink = None
    action = "info"
    boot = "cold"
    filename = None 
    targets = None  
    bl = None  

    exit_result = 0

    if len(sys.argv) < 2:
        print()
        print("==============================")
        print(" CrazyLoader Flash Utility")
        print("==============================")
        print()
        print(" Usage:", sys.argv[0], "[CRTP options] <action> [parameters]")
        print()
        print("The CRTP options are described above")
        print()
        print("Crazyload option:")
        print("   info                    : Print the info of the bootloader "
              "and quit.")
        print("                             Will let the target in bootloader "
              "mode")
        print("   reset                   : Reset the device in firmware mode")
        print("   flash <file> [targets]  : flash the <img> binary file from "
              "the first")
        print("                             possible  page in flash and reset "
              "to firmware")
        print("                             mode.")
        sys.exit(0)

    # Analyse the command line parameters
    sys.argv = sys.argv[1:]
    argv = []

    i = 0
    while i < len(sys.argv):
        if sys.argv[i] == "--cold-boot" or sys.argv[i] == "-c":
            boot = "cold"
        elif sys.argv[i] == "--warm-boot" or sys.argv[i] == "-w":
            boot = "reset"
            i += 1
            clink = sys.argv[i]
        else:
            argv += [sys.argv[i]]
        i += 1
    sys.argv = argv

    # Analyse the command
    if len(sys.argv) < 1:
        action = "info"
    elif sys.argv[0] == "info":
        action = "info"
    elif sys.argv[0] == "reset":
        action = "reset"
    elif sys.argv[0] == "flash":
        if len(sys.argv) < 2:
            print("The flash action require a file name.")
            sys.exit(-1)
        action = "flash"
        filename = sys.argv[1]
        targets = []  # Dict[Target]
        for t in sys.argv[2:]:
            if t.startswith("deck-"):
                [deck, target, type] = t.split("-")
                targets.append(Target("deck", target, type, [], []))
            else:
                [target, type] = t.split("-")
                targets.append(Target("cf2", target, type, [], []))
    else:
        print("Action", sys.argv[0], "unknown!")
        sys.exit(-1)

    try:
        # Initialise the bootloader lib
        bl = Bootloader(clink)

        warm_boot = (boot == "reset")
        if warm_boot:
            print("Reset to bootloader mode ...")
            sys.stdout.flush()
        else:  # The connection is done by a cold boot ...
            print("Restart the Crazyflie you want to bootload in the next"),
            print(" 10 seconds ..."),

            sys.stdout.flush()

        ######################################
        # Doing something (hopefully) useful
        ######################################

        if action == "info":
            def print_info(version: int, connected_targets: [Target]):
                print("Connected to bootloader on {} (version=0x{:X})".format(
                    BootVersion.to_ver_string(version),
                    version
                    )
                )
                for target in connected_targets:
                    print(target)

            # flash_full called with no filename will not flash, just call
            # our info callback
            bl.flash_full(None, None, warm_boot, None, print_info)
        elif action == "flash" and filename:
            is_target_required = not filename.endswith('.zip')
            if (is_target_required and not targets):
                print("The flash action with a non .zip file requires a target")
                sys.exit(-1)
            try:
                bl.flash_full(None, filename, warm_boot, targets)
            except Exception as e:
                print("Failed to flash: {}".format(e))
                exit_result = -1
        elif action == "reset":
            bl.reset_to_firmware()
        else:
            None
    except Exception as e:
        import traceback

        traceback.print_exc(file=sys.stdout)
        print(e)
        exit_result = -1

    finally:
        #########################
        # Closing the connection
        #########################
        if bl:
            bl.close()

    sys.exit(exit_result)

**Reply by knmcguire** [reply]
I see that you are trying to run this on a windows computer with no docker installed as according to the instructions on top of the getting started guide. Try to find a computer with ubuntu installed or perhaps use a VM.

**Reply by akshaysunkara** [reply]
Hey! I just tried using a VM to run docker on. 

This is what I get whenever I try running the code:

![image](https://github.com/bitcraze/discussions/assets/171630750/ea04b9bd-9f1f-4d22-bdda-339b498d7fdc)

I'm connected to the drone on the crazyflie GUI and connected my AI deck and my ARN-JTAG 20-10 adapter.

Just wondering what the issue meant and how I could fix it. Does this mean my AI deck isn't being detected?

Also whenever I flash it in my VM, all it does is make the drone restart but I think it's because I've flashed it before outside of the VM. 

Thanks a lot again.

**Reply by akshaysunkara** [reply]
![Untitled](https://github.com/bitcraze/discussions/assets/171630750/2d5cb2cb-e776-4cf0-91f1-50ee808945eb)
![Untitled-1](https://github.com/bitcraze/discussions/assets/171630750/68de3dd3-e445-46da-9992-6b10c881800c)

Idk if the problem is that I don't have the drone connected VIA Usb. I have it connected via crazyradio on the GUI. I'm not able to connect it through USB on a VM because it says that it's "busy with another request."

**Reply by gemenerik** [reply]
It looks like you have not connected your debugger to your pc. You do not need to connect to the main board of the CF. Just make sure the ai-deck is powered. If you use a different debugger than the default we recommend (arm-usb-tiny-h) you will have change some defaults. Note the `export GAPY_OPENOCD_CABLE=interface/ftdi/olimex-arm-usb-tiny-h.cfg` in your given command.

**Reply by akshaysunkara** [reply]
Hey, I'm trying to use the debugger right now.

I have it connected to my PC and I'm using the VMWARE and I get the error that the burst read has failed. Should I connect the debugger to anything or should the only thing connected to my laptop by my Crazyradio and the debugger?

**Reply by akshaysunkara** [reply]
The crazyflie drone is turned on and the AI deck is powered

**Reply by akshaysunkara** [reply]
Flashing image with command:
gap8-openocd -d0 -c "gdb_port disabled; telnet_port disabled; tcl_port disabled" -c "script interface/ftdi/olimex-arm-usb-tiny-h.cfg; script target/gap8revb.tcl; script tcl/flash_image.tcl; script tcl/jtag_boot.tcl; gap_flash_raw_hyper /module/BUILD/GAP8_V2/GCC_RISCV_FREERTOS/target.board.devices.flash.img 76832 /gap_sdk/utils/openocd_tools; exit;"
Open On-Chip Debugger 0.10.0+dev-gb84d97ec4 (2022-04-20-14:26)
Licensed under GNU GPL v2
For bug reports, read
http://openocd.org/doc/doxygen/bugs.html
debug_level: 0

DEPRECATED! use 'adapter speed' not 'adapter_khz'
DEPRECATED! use 'adapter driver' not 'interface'
TARGET create
GAP8 INIT TARGET
Error: JTAG scan chain interrogation failed: all ones
Error: Check JTAG interface, timings, target power, etc.
Error: Trying to use configured scan chain anyway...
Error: gap8.cpu: IR capture error; saw 0x0f not 0x01
GAP8 examine target
Init jtag
Initialising GAP8 JTAG TAP
Error: Burst read failed
Error: Burst read failed
^CError: Burst read failed
can't read halt state
Error: Couldn't read the CPU state

This is the errors that I'm getting

**Reply by akshaysunkara** [reply]
I connected the ARM-JTAG 20-10 adapter to the drone like I should've and my GAP8 LED on my AI deck blinks green periodically. The CPU state error I got was likely due to me disconnecting the ARM-USB-TINY-H but its still not failing to burst read. 

Anything that I'm doing wrong?

**Reply by knmcguire** [reply]
Hi, Sorry but Rik is now on a holiday so he should be back in 2 weeks.

In terms of the flashing, this seems like multiple things are going wrong, maybe the connecter was mounted upside down? do you have another lint cable to try out?

**Reply by gemenerik** [reply]
In the images above it looks like your JTAG link cable is upside down, please [connect as shown here](https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/infrastructure/flashing/). Note the red wire. Please re-follow the instructions carefully, including the detailed flashing instructions, and let me know if it works.
